src/main/java/com/zendesk/maxwell/row/RowIdentity.java

```diff
@@ -0,0 +1,107 @@
+package com.zendesk.maxwell.row;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.*;
+
+public class RowIdentity implements Serializable {
+	private String database;
+	private String table;
+	private final List<Map.Entry<String, Object>> primaryKeyColumns;
+
+	// slightly less verbose pair construction
+	public static Map.Entry<String, Object> pair(String key, Object value) {
+		return new AbstractMap.SimpleImmutableEntry<>(key, value);
+	}
+
+	public RowIdentity(String database, String table, List<Map.Entry<String, Object>> primaryKeyColumns) {
+		this.database = database;
+		this.table = table;
+		this.primaryKeyColumns = primaryKeyColumns;
+	}
+
+	public String getDatabase() {
+		return database;
+	}
+
+	public String getTable() {
+		return table;
+	}
+
+	public String toKeyJson(RowMap.KeyFormat keyFormat) throws IOException {
+		MaxwellJson json = MaxwellJson.getInstance();
+		JsonGenerator g = json.reset();
+		if ( keyFormat == RowMap.KeyFormat.HASH )
+			writeKeyJsonHash(g);
+		else
+			writeKeyJsonArray(g);
+		return json.consume();
+	}
+
+	private void writeKeyJsonHash(JsonGenerator g) throws IOException {
+		writeStartCommon(g);
+		if (primaryKeyColumns.isEmpty()) {
+			g.writeStringField(FieldNames.UUID, UUID.randomUUID().toString());
+		} else {
+			for (Map.Entry<String,Object> pk : primaryKeyColumns) {
+				writePrimaryKey(g, "pk." + pk.getKey().toLowerCase(), pk.getValue());
+			}
+		}
+		g.writeEndObject();
+	}
+
+	private void writeKeyJsonArray(JsonGenerator g) throws IOException {
+		g.writeStartArray();
+		g.writeString(database);
+		g.writeString(table);
+
+		g.writeStartArray();
+		for (Map.Entry<String,Object> pk : primaryKeyColumns) {
+			g.writeStartObject();
+			MaxwellJson.writeValueToJSON(g, true, pk.getKey().toLowerCase(), pk.getValue());
+			g.writeEndObject();
+		}
+		g.writeEndArray();
+		g.writeEndArray();
+	}
+
+	public String toConcatString() {
```

Comment:
while you're in here let's add some minimal documentation to these functions, just explain the format each is outputting and where it's used...

Reply:
Sure, done in e63680bcc76ffdc17f70664316d28dcc9536259b